optax/_src/loss_test.py

```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
+        1, nclasses, size=(batchsize, steps)).astype(np.int32)
+
+    # This function only covers the cases without same-label repetition.
+    # `test_repeat_with_one_to_one_alignment` below complements those cases.
+    # So, redraw the samples for satisfying the non-repetition constraint.
+    for n in range(labels.shape[0]):
+      for t in range(1, labels.shape[1]):
+        while labels[n, t] == labels[n, t - 1]:
+          labels[n, t] = np.random.uniform(1, nclasses)
+
+    per_seq_loss = self.variant(loss.ctc_loss)(
+        logits, np.zeros(logits.shape[:2]), labels, np.zeros(labels.shape))
+
+    logprobs = jax.nn.log_softmax(logits)
+    for b in range(batchsize):
+      p = 0.0
+      for t in range(steps):
+        p += logprobs[b, t, labels[b, t]]
+      np.testing.assert_allclose(jnp.array(-p), per_seq_loss[b],
+                                 rtol=1e-6)
+
+  @chex.all_variants
+  def test_with_one_to_one_alignment_and_paddings(self):
+    # test if padding indicators work as expected.
+    batch_size = 1
+    nclasses = 8
+    steps = 4
+    logits = np.random.normal(size=[batch_size, steps, nclasses])
+    logprobs = jax.nn.log_softmax(logits)
+    labels = np.array([[1, 2, 3, 4]])
+    logit_padding = np.array([[0, 0, 0, 1]], dtype=np.float32)
+    label_padding = np.array([[0, 0, 0, 1]], dtype=np.float32)
+
+    value_and_grad = self.variant(jax.value_and_grad(_average_ctc_loss))
+    actual_loss, actual_gradients = value_and_grad(
+        logits, logit_padding, labels, label_padding)
+    expected_loss = -(logprobs[0, 0, 1] + logprobs[0, 1, 2] + logprobs[0, 2, 3])
+    np.testing.assert_allclose(expected_loss, actual_loss)
+
+    expected_gradients = np.array(jax.nn.softmax(logits))
+    expected_gradients[0, 0, 1] -= 1.0
+    expected_gradients[0, 1, 2] -= 1.0
+    expected_gradients[0, 2, 3] -= 1.0
+    expected_gradients[0, 3, :] = 0.0
+    np.testing.assert_allclose(expected_gradients, actual_gradients,
+                               rtol=1e-6)
+
+  @chex.all_variants
+  def test_repeat_with_one_to_one_alignment(self):
+    # test if it can correctly handle the same-label repetition.
+    batch_size = 2
+    labels = np.array([
+        [1, 2, 2, 3],
+        [2, 3, 4, 4],
+    ])
+    label_lens = np.array([4, 4])
+    label_paddings = _lengths_to_paddings(label_lens, 4)
+    logits = np.random.randn(batch_size, 5, 5)
+    logprobs = jax.nn.log_softmax(logits)
+    logprob_paddings = np.zeros(logprobs.shape[:2])
+
+    per_seq_loss = self.variant(loss.ctc_loss)(
+        logprobs, logprob_paddings, labels, label_paddings)
+
+    expected_alignment = [
+        [1, 2, 0, 2, 3],
+        [2, 3, 4, 0, 4],
+    ]
+
+    for n in range(batch_size):
+      expected_loss = -sum(logprobs[n, t, k]
+                           for t, k in enumerate(expected_alignment[n]))
+      np.testing.assert_allclose(
+          jnp.array(expected_loss), per_seq_loss[n], rtol=1e-6)
+
 if __name__ == '__main__':
```

Comment:
optional: consider add a test comparing your implementation w/ TF on both the forward pass and backward pass w various values. the tolerance of the test won't be very good though, so it might not be super helpful.

Reply:
I was just reluctant to add another test dependency package.  Actually, I have such tests already, and if it's okay to add test dependency I can move the code to this repo.

Reply:
It is a good suggestion! 
We should consider adding parity tests against TF implementations; I'll discuss it with other Optax core devs. In any case, this is out of scope of this PR.